golang/src/user-preferences/main.go

```diff
@@ -0,0 +1,393 @@
+              FROM user_preferences p,
+                   users u
+             WHERE p.user_id = u.id
+               AND u.username = $1`
+	rows, err := u.db.Query(query, username)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var prefs []UserPreferencesRecord
+	for rows.Next() {
+		var pref UserPreferencesRecord
+		if err := rows.Scan(&pref.ID, &pref.UserID, &pref.Preferences); err != nil {
+			return nil, err
+		}
+		prefs = append(prefs, pref)
+	}
+	if err := rows.Err(); err != nil {
+		return prefs, err
+	}
+	return prefs, nil
+}
+
+// insertPreferences adds a new preferences to the database for the user.
+func (u *UserPreferencesApp) insertPreferences(username, prefs string) error {
+	query := `INSERT INTO user_preferences (user_id, preferences)
+                 VALUES ($1, $2)`
+	userID, err := queries.UserID(u.db, username)
+	if err != nil {
+		return err
+	}
+	_, err = u.db.Exec(query, userID, prefs)
+	return err
+}
+
+// updatePreferences updates the preferences in the database for the user.
+func (u *UserPreferencesApp) updatePreferences(username, prefs string) error {
+	query := `UPDATE ONLY user_preferences
+                    SET preferences = $2
+                  WHERE user_id = $1`
+	userID, err := queries.UserID(u.db, username)
+	if err != nil {
+		return err
+	}
+	_, err = u.db.Exec(query, userID, prefs)
+	return err
+}
+
+// deletePreferences deletes the user's preferences from the database.
+func (u *UserPreferencesApp) deletePreferences(username string) error {
+	query := `DELETE FROM ONLY user_preferences WHERE user_id = $1`
+	userID, err := queries.UserID(u.db, username)
+	if err != nil {
+		return err
+	}
+	_, err = u.db.Exec(query, userID)
+	return err
+}
+
+// Greeting prints out a greeting to the writer.
+func (u *UserPreferencesApp) Greeting(writer http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(writer, "Hello from user-preferences.")
+}
+
+func badRequest(writer http.ResponseWriter, msg string) {
+	writer.WriteHeader(http.StatusBadRequest)
+	writer.Write([]byte(msg))
+	logcabin.Error.Print(msg)
+}
+
+func errored(writer http.ResponseWriter, msg string) {
+	writer.WriteHeader(http.StatusInternalServerError)
+	writer.Write([]byte(msg))
+	logcabin.Error.Print(msg)
+}
+
+func (u *UserPreferencesApp) getUserPreferencesForRequest(username string, wrap bool) ([]byte, error) {
+	prefs, err := u.getPreferences(username)
+	if err != nil {
+		return nil, fmt.Errorf("Error getting preferences for username %s: %s", username, err)
+	}
+	var retval UserPreferencesRecord
+	if len(prefs) >= 1 {
+		retval = prefs[0]
+	}
+	response, err := convert(&retval, wrap)
+	if err != nil {
+		return nil, fmt.Errorf("Error generating response for username %s: %s", username, err)
+	}
+	var jsoned []byte
+	if len(response) > 0 {
+		jsoned, err = json.Marshal(response)
+		if err != nil {
+			return nil, fmt.Errorf("Error generating session JSON for user %s: %s", username, err)
```

Comment:
(okay maybe not completely stopped, but this one's in actual code rather than just a comment)